jepsen/test/jepsen/tests/causal_reverse_test.clj

```diff
@@ -0,0 +1,63 @@
+(ns jepsen.tests.causal-reverse-test
+  (:require [jepsen.tests.causal-reverse :refer :all]
+            [knossos.op :refer :all]
+            [clojure.test :refer :all]
+            [jepsen.checker :as checker]))
+
+(deftest casusal-reverse-test
+  (testing "Can validate sequential histories"
+    (let [c (checker)
+          valid [(invoke 0 :write 1)
+                 (ok     0 :write 1)
+                 (invoke 0 :write 2)
+                 (ok     0 :write 2)
+                 (invoke 0 :read nil)
+                 (ok     0 :read [1 2])]
+          two-without-one [(invoke 0 :write 1)
+                           (ok     0 :write 1)
+                           (invoke 0 :write 2)
+                           (ok     0 :write 2)
+                           (invoke 0 :read nil)
+                           (ok     0 :read [2])]
+          bigger [(invoke 0 :write 1)
+                  (ok     0 :write 1)
+                  (invoke 0 :write 2)
+                  (ok     0 :write 2)
+                  (invoke 0 :write 3)
+                  (ok     0 :write 3)
+                  (invoke 0 :write 4)
+                  (ok     0 :write 4)
+                  (invoke 0 :write 5)
+                  (ok     0 :write 5)
+                  (invoke 0 :read nil)
+                  (ok     0 :read [1 2 3 4 5])]]
+      (is (:valid?      (checker/check c nil valid nil)))
+      (is (not (:valid? (checker/check c nil two-without-one nil))))
+      (is (:valid?      (checker/check c nil bigger nil)))))
+
+  (testing "Can validate concurrent histories"
+    (let [c (checker)
+          concurrent1 [(invoke 0 :write 2)
+                       (invoke 0 :write 1)
+                       (ok     0 :write 1)
+                       (invoke 0 :read nil)
+                       (ok     0 :write 2)
+                       (ok     0 :read [1 2])]
+          concurrent2  [(invoke 0 :write 1)
+                        (invoke 0 :write 2)
+                        (ok     0 :write 1)
+                        (invoke 0 :read nil)
+                        (ok     0 :write 2)
+                        (ok     0 :read [2 1])]]
+      (is (:valid? (checker/check (checker) nil concurrent1 nil)))
+      (is (:valid? (checker/check (checker) nil concurrent2 nil)))))
+
+  (testing "Can detect reverse causal anomaly"
+    (let [c (checker)
+          reverse-causal-read [(invoke 0 :write 1)
+                               (ok     0 :write 1)
+                               (invoke 0 :write 2)
+                               (ok     0 :write 2)
+                               (invoke 0 :read nil)
+                               (ok     0 :read [2 1])]]
```

Comment:
The failing test case

Reply:
Huh, yeah, that does look like a problem! It's been a couple years since I wrote this; maybe I have the test case wrong, or the checker is incorrect.

Reply:
Aha, yes, I misremembered what this checker actually does. We only verify that completed writes appear before later ones, but we don't verify the order they appear in. A valid read has to have both 1 and 2, but in any order. Invalid reads would be [1] or [2], I think.